lock-api/src/main/java/com/palantir/lock/client/TransactionStarter.java

```diff
@@ -138,16 +131,20 @@ public void close() {
     }
 
     @VisibleForTesting
-    static Consumer<List<BatchElement<Void, StartIdentifiedAtlasDbTransactionResponse>>> consumer(
+    static Consumer<List<BatchElement<Integer, List<StartIdentifiedAtlasDbTransactionResponse>>>> consumer(
             LockLeaseService lockLeaseService, LockWatchEventCache lockWatchEventCache) {
         return batch -> {
-            int numTransactions = batch.size();
+            int numTransactions = batch.stream().map(BatchElement::argument).reduce(0, Integer::sum);
 
             List<StartIdentifiedAtlasDbTransactionResponse> startTransactionResponses =
                     getStartTransactionResponses(lockLeaseService, lockWatchEventCache, numTransactions);
 
-            for (int i = 0; i < numTransactions; i++) {
-                batch.get(i).result().set(startTransactionResponses.get(i));
+            int start = 0;
+            for (BatchElement<Integer, List<StartIdentifiedAtlasDbTransactionResponse>> batchElement
+                    : batch) {
+                int end = start + batchElement.argument();
+                batchElement.result().set(startTransactionResponses.subList(start, end));
```

Comment:
Hmm, just think very carefully about what this subList does here. @jeremyk-91

Reply:
Yes, since all of the returned lists are a view of the original list `startTransactionResponses`, we can get `ConcurrentModificationException`s or other such things if the lists returned are abused (i.e. elements are removed). I don't think this will be a problem given that we don't do that, but we could change this to return a fresh list for each result if it is a concern.

Reply:
@Jolyon-S The tricky thing here is that subList keeps a reference to the original list, so the entire response will remain live in terms of memory as long as any consumer keeps a reference to their chunk of the list. This _looks_ like it would probably be safe from tracing through the code, but I'd prefer to just create a new list in case there's something silly around threads contending over `updateAndGet()` or things like that.